src/app/marketplace/edit-ad.component.ts

```diff
@@ -0,0 +1,329 @@
+import {
+  AdDataForEdit, AdKind, AdDataForNew, Image, Currency,
+  AdBasicData, AdCategoryWithChildren, AdEdit
+} from 'app/api/models';
+import { OnInit, ChangeDetectionStrategy, Component, Injector, ChangeDetectorRef } from '@angular/core';
+import { BasePageComponent } from 'app/shared/base-page.component';
+import { Menu } from 'app/shared/menu';
+import { MarketplaceService, ImagesService } from 'app/api/services';
+import { MarketplaceHelperService } from 'app/core/marketplace-helper.service';
+import { Observable, BehaviorSubject } from 'rxjs';
+import { FormGroup, Validators } from '@angular/forms';
+import { ManageImagesComponent } from 'app/shared/manage-images.component';
+import { take } from 'rxjs/operators';
+import { empty, validateBeforeSubmit } from 'app/shared/helper';
+import { BsModalService } from 'ngx-bootstrap/modal';
+
+type HierarchyItem = AdCategoryWithChildren & {
+  level: number,
+  leaf: boolean
+};
+
+const IMAGE_MANAGED_TIMEOUT = 6_000;
+
+/**
+ * Edits an advertisement or webshop
+ */
+@Component({
+  selector: 'edit-ad',
+  templateUrl: 'edit-ad.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class EditAdComponent
+  extends BasePageComponent<AdDataForNew | AdDataForEdit>
+  implements OnInit {
+
+  static BASED_ON_ID: string;
```

Comment:
I didn't like this way of passing the parameter. Instead of using a hacky static variable, try using query parameters. See below for the comment on line 198.